mapper/thin_edge_json/src/lib.rs

```diff
@@ -0,0 +1,574 @@
+
+    fn new_invalid_utf8(bytes: &[u8], from: std::str::Utf8Error) -> ThinEdgeJsonError {
+        let index = from.valid_up_to();
+        let input = std::str::from_utf8(&bytes[..index]).unwrap_or("");
+
+        ThinEdgeJsonError::InvalidUTF8 {
+            input_excerpt: input_prefix(input, ThinEdgeJsonError::MAX_LEN),
+            from,
+        }
+    }
+
+    fn new_invalid_json(input: &str, from: json::JsonError) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidJson {
+            input_excerpt: input_prefix(input, ThinEdgeJsonError::MAX_LEN),
+            from,
+        }
+    }
+
+    fn new_invalid_json_root(json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonRoot {
+            json_excerpt: input_prefix(&json.to_string(), ThinEdgeJsonError::MAX_LEN),
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn new_invalid_json_value(name: &str, json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonValue {
+            name: String::from(name),
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn new_invalid_json_time(json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonTime {
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn json_type(input: &JsonValue) -> &'static str {
+        match input {
+            JsonValue::String(_) | JsonValue::Short(_) => "a string",
+            JsonValue::Number(_) => "a number",
+            JsonValue::Object(_) => "an object",
+            JsonValue::Array(_) => "an array",
+            JsonValue::Boolean(_) => "a boolean",
+            JsonValue::Null => "null",
+        }
+    }
+}
+
+#[cfg(test)]
+#[macro_use]
+extern crate pretty_assertions;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn thin_edge_json_reject_invalid_utf8() {
+        let input = b"temperature\xc3\x28";
+
+        let expected_error =
+            r#"Invalid UTF8: invalid utf-8 sequence of 1 bytes from index 11: temperature..."#;
+        let output = ThinEdgeJson::read_from_bytes(input);
+
+        let error = output.unwrap_err();
+        assert_eq!(expected_error, error.to_string());
+    }
+
+    #[test]
+    fn thin_edge_json_reject_non_utf8_input() {
+        let input = b"\xc3\x28";
+
+        let expected_error = r#"Invalid UTF8: invalid utf-8 sequence of 1 bytes from index 0: ..."#;
+        let output = ThinEdgeJson::read_from_bytes(input);
+
+        let error = output.unwrap_err();
+        assert_eq!(expected_error, error.to_string());
+    }
+
+    #[test]
+    fn thin_edge_json_reject_arrays() {
+        let input = r"[50,23]";
+
+        let expected_error = r#"Invalid Thin Edge measurement: it cannot be an array: [50,23]"#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```

Comment:
This test should use `ThinEdgeJson::from_str()`.

1. You expect an error, so you don't need this weird `ThinEdgeJson::read_from_bytes` function.
2. You have a string as input and a function to read a string. There is no point to build a temp `[u8]`.